src/graphics/mesh3d.rs

```diff
@@ -0,0 +1,336 @@
+    #[inline]
+    pub fn from_min_max(minimum: Vec3, maximum: Vec3) -> Self {
+        let minimum = minimum;
+        let maximum = maximum;
+        let center = 0.5 * (maximum + minimum);
+        let half_extents = 0.5 * (maximum - minimum);
+        Self {
+            center: center.into(),
+            half_extents: half_extents.into(),
+        }
+    }
+}
+
+/// Transform3d is used to transform 3d objects.
+#[derive(Debug, Copy, Clone)]
+pub struct Transform3d {
+    /// The position or translation of this `Transform3d`
+    pub position: mint::Vector3<f32>,
+    /// The rotation of this `Transform3d`
+    pub rotation: mint::Quaternion<f32>,
+    /// The scale of this `Transform3d`
+    pub scale: mint::Vector3<f32>,
+}
+
+impl Default for Transform3d {
+    fn default() -> Self {
+        Self {
+            position: Vec3::new(0.0, 0.0, 0.0).into(),
+            rotation: glam::Quat::IDENTITY.into(),
+            scale: Vec3::new(1.0, 1.0, 1.0).into(),
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, bytemuck::Pod, bytemuck::Zeroable)]
+pub(crate) struct Instance3d {
+    transform: [[f32; 4]; 4],
+    color: [f32; 4],
+}
+
+impl Default for Instance3d {
+    fn default() -> Self {
+        Self::from_param(&DrawParam3d::default(), Vec3::ZERO)
+    }
+}
+
+impl Instance3d {
+    pub(crate) fn desc() -> wgpu::VertexBufferLayout<'static> {
+        use std::mem;
+        wgpu::VertexBufferLayout {
+            array_stride: mem::size_of::<Instance3d>() as wgpu::BufferAddress,
+            step_mode: wgpu::VertexStepMode::Instance,
+            attributes: &[
+                wgpu::VertexAttribute {
+                    offset: 0,
+                    shader_location: 5,
+                    format: wgpu::VertexFormat::Float32x4,
+                },
+                // A mat4 takes up 4 vertex slots as it is technically 4 vec4s. We need to define a slot
+                // for each vec4. We don't have to do this in code though.
+                wgpu::VertexAttribute {
+                    offset: mem::size_of::<[f32; 4]>() as wgpu::BufferAddress,
+                    shader_location: 6,
+                    format: wgpu::VertexFormat::Float32x4,
+                },
+                wgpu::VertexAttribute {
+                    offset: mem::size_of::<[f32; 8]>() as wgpu::BufferAddress,
+                    shader_location: 7,
+                    format: wgpu::VertexFormat::Float32x4,
+                },
+                wgpu::VertexAttribute {
+                    offset: mem::size_of::<[f32; 12]>() as wgpu::BufferAddress,
+                    shader_location: 8,
+                    format: wgpu::VertexFormat::Float32x4,
+                },
+                wgpu::VertexAttribute {
+                    offset: mem::size_of::<[f32; 16]>() as wgpu::BufferAddress,
+                    shader_location: 9,
+                    format: wgpu::VertexFormat::Float32x4,
+                },
+            ],
```

Comment:
You could use [`wgpu::vertex_attr_array!`](https://docs.rs/wgpu/latest/wgpu/macro.vertex_attr_array.html) here

Reply:
Fixed in latest commit